File: Trees/Python/Path Sum.py
# Problem link: https://leetcode.com/problems/path-sum/description/
# Time: O(n)
# Space: O(n)
class Solution:
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
        return self.dfs(root, targetSum, 0)
    
    def dfs(self, root, targetSum, preSum):
        if not root:
            return False

        preSum += root.val
        
        if not root.left and not root.right and preSum == targetSum:
            return True
        
        return self.dfs(root.left, targetSum, preSum) or self.dfs(root.right, targetSum, preSum)

File: Trees/Java/Path Sum.java
// Problem link: https://leetcode.com/problems/path-sum/description/
// Time: O(n)
// Space: O(n)
class Solution {
    public boolean hasPathSum(TreeNode root, int targetSum) {
        return this.dfs(root, targetSum, 0);
    }

    private boolean dfs(TreeNode root, int targetSum, int preSum) {
        if(root == null) {
            return false;
        }

        preSum += root.val;

        if(preSum == targetSum && root.left == null && root.right == null) {
            return true;
        }

        return dfs(root.left, targetSum, preSum) || dfs(root.right, targetSum, preSum);
    }
}
